interfaces/builtin/pkcs11.go

```diff
@@ -0,0 +1,160 @@
+
+// Name of the pkcs11 interface.
+func (iface *pkcs11Interface) Name() string {
+	return "pkcs11"
+}
+
+func (iface *pkcs11Interface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              pkcs11Summary,
+		BaseDeclarationSlots: pkcs11BaseDeclarationSlots,
+	}
+}
+
+func (iface *pkcs11Interface) String() string {
+	return iface.Name()
+}
+
+const pkcs11PermanentSlotSecComp = `
+# Description: Allow operating as an p11-kit server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+
+# Needed by server upon client connect
+accept
+accept4
+`
+
+func (iface *pkcs11Interface) getSocketPath(slot *snap.SlotInfo) (string, error) {
+	socketAttr, isSet := slot.Attrs["pkcs11-socket"]
+	if !isSet {
+		return "", fmt.Errorf(`cannot use pkcs11 slot %s without "pkcs11-socket" attribute`, slot.Name)
+	}
+
+	socketPath, ok := socketAttr.(string)
+	if !ok {
+		return "", fmt.Errorf(`pkcs11: "pkcs11-socket" attribute must be a string, not %v`,
+			slot.Attrs["pkcs11-socket"])
+	}
+
+	if ok := cleanSubPath(socketPath); !ok {
+		return "", fmt.Errorf("slot %q, a unix socket path is not clean", slot.Name)
+	}
+
+	// separate socket name and check socket is in /run/p11-kit
+	if filepath.Dir(socketPath) != "/run/p11-kit" {
+		return "", fmt.Errorf("slot %q, a unix socket has to be in /run/p11-kit directory", slot.Name)
+	}
+
+	return socketPath, nil
+}
+
+func (iface *pkcs11Interface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	socketPath, err := iface.getSocketPath(slot)
+	if err != nil {
+		return err
+	}
+
+	if err := apparmor_sandbox.ValidateNoAppArmorRegexp(socketPath); err != nil {
+		return fmt.Errorf("pkcs11 interface socket path is invalid: %v", err)
+	}
+	return err
+}
+
+func (iface *pkcs11Interface) SecCompPermanentSlot(spec *seccomp.Specification, slot *snap.SlotInfo) error {
+	spec.AddSnippet(pkcs11PermanentSlotSecComp)
+	return nil
+}
+
+func (iface *pkcs11Interface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var socketPath string
+	if err := slot.Attr("pkcs11-socket", &socketPath); err != nil || socketPath == "" {
+		return fmt.Errorf("slot %q must have a unix socket 'pkcs11-socket' attribute", slot.Ref())
+	}
+
+	socketRule := fmt.Sprintf(`"/{,var/}%s" rw,`, socketPath[1:])
+	// pkcs11 server unix socket
+	spec.AddSnippet(fmt.Sprintf(`# pkcs11 socket
+%s
+# pkcs11 config for p11-proxy
+/etc/pkcs11/{,**} r,
+# pkcs11 tools
+/usr/bin/p11tool ixr,
+/usr/bin/pkcs11-tool ixr,`,
+		socketRule))
+	return nil
+}
+
+func (iface *pkcs11Interface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	socketPath, err := iface.getSocketPath(slot)
+	if err != nil {
+		return err
+	}
+
+	socketRule := fmt.Sprintf(`"/{,var/}%s" rwk,`, socketPath[1:])
+	spec.AddSnippet(fmt.Sprintf(`# pkcs11 socket dir
+/{,var/}run/p11-kit/  rw,
+%s
+# pkcs11 config
+/etc/pkcs11/{,**} r,
```

Comment:
What snap is expected to provide these binaries / config for pkcs11? I can't see anything in the current base core snaps.